meteor/iterative.py

```diff
@@ -0,0 +1,270 @@
+        then return
+
+    max_iterations: int
+        If this number of iterations is reached, stop early.
+
+    tv_weights_to_scan : list[float], optional
+        A list of TV regularization weights (λ values) to be scanned for optimal results,
+        by default [0.001, 0.01, 0.1, 1.0].
+
+    Returns
+    -------
+    output_dataset: rs.DataSet
+        The estimated derivative phases, along with the input amplitudes and input computed phases.
+
+    metadata: pd.DataFrame
+        Information about the algorithm run as a function of iteration. For each step, includes:
+        the tv_weight used, the negentropy (after the TV step), and the average phase change in
+        degrees.
+    """
+
+    # clean TV denoising interface that is crystallographically intelligent
+    # maintains state for the HKL index, spacegroup, and cell information
+    def tv_denoise_closure(difference: np.ndarray) -> tuple[np.ndarray, TvDenoiseResult]:
+        delta_amp, delta_phase = complex_array_to_rs_dataseries(
+            difference, index=input_dataset.index
+        )
+
+        # these two names are only used inside this closure
+        delta_amp.name = "DF_for_tv_closure"
+        delta_phase.name = "DPHI_for_tv_closure"
+
+        diffmap = rs.concat([delta_amp, delta_phase], axis=1)
+        diffmap.cell = input_dataset.cell
+        diffmap.spacegroup = input_dataset.spacegroup
+
+        denoised_map_coefficients, tv_metadata = tv_denoise_difference_map(
+            diffmap,
+            difference_map_amplitude_column=delta_amp.name,
+            difference_map_phase_column=delta_phase.name,
+            lambda_values_to_scan=tv_weights_to_scan,
+            full_output=True,
+        )
+
+        denoised_difference = rs_dataseies_to_complex_array(
+            denoised_map_coefficients[delta_amp.name], denoised_map_coefficients[delta_phase.name]
+        )
+
+        return denoised_difference, tv_metadata
+
+    # convert the native and derivative datasets to complex arrays
+    native = rs_dataseies_to_complex_array(
+        input_dataset[native_amplitude_column], input_dataset[calculated_phase_column]
+    )
+    initial_derivative = rs_dataseies_to_complex_array(
+        input_dataset[derivative_amplitude_column], input_dataset[calculated_phase_column]
+    )
+
+    # estimate the derivative phases using the iterative TV algorithm
+    it_tv_complex_derivative, metadata = _complex_derivative_from_iterative_tv(
+        native=native,
+        initial_derivative=initial_derivative,
+        tv_denoise_function=tv_denoise_closure,
+        convergence_tolerance=convergence_tolerance,
+        max_iterations=max_iterations,
+    )
+    _, derivative_phases = complex_array_to_rs_dataseries(
+        it_tv_complex_derivative, input_dataset.index
+    )
+
+    # combine the determined derivative phases with the input to generate a complete output
+    output_dataset = input_dataset.copy()
```

Comment:
do we want to copy or just add a column to the og dataset? From our discussions I thought we have converged to adding a column

Reply:
discussed in person, conclusion -- return copy with new column in these situations